### PR TITLE
Generate: length validation

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

This PR moves the length validation code to a dedicated function, to make `generate` more readable, and adds code to handle two cases discussed recently in other issues:
1. `min_new_tokens` was not being checked against the maximum possible generation length;
2. users were unaware that a default `max_length` exists, which impacts `min_length` and `min_new_tokens`

### CODE REVIEW

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -1245,6 +1245,46 @@ def _validate_model_kwargs(self, model_kwargs: Dict[str, Any]):
...
+                f" {generation_config.max_length}. This can lead to unexpected behavior. You should consider"
+                " increasing `max_new_tokens`."
+            )
+
+        # 2. Min length exceptions due to unfeasible parameter combinations
+        min_length_error_suffix = " To continue, decrease the minimum length and/or increase the maximum length."
+        if has_default_max_length:
+            min_length_error_suffix += (
+                f" Note that `max_length` is set to {generation_config.max_length}, its default value."
+            )
```
Comment (gante) [inline src/transformers/generation/utils.py:1273]: This should make it clear when any minimum length argument clashes with the default value (whose existence is often unknown by the users)

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -1245,6 +1245,46 @@ def _validate_model_kwargs(self, model_kwargs: Dict[str, Any]):
...
+                f" the maximum possible length ({generation_config.max_length})." + min_length_error_suffix
+            )
+        if generation_config.min_new_tokens is not None:
+            min_length = generation_config.min_new_tokens + input_ids_length
+            if min_length > generation_config.max_length:
+                raise ValueError(
+                    f"Unfeasible length constraints: `min_new_tokens` ({generation_config.min_new_tokens}), when "
+                    f"added to the prompt length ({input_ids_length}), is larger than"
+                    f" the maximum possible length ({generation_config.max_length})." + min_length_error_suffix
+                )
```
[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/generation/utils.py:1286]: Should this be a warning with `min_new_tokens` decreased to the max possible value?

**Review by sgugger — APPROVED** [review]
Thanks for the cleanup!

File: src/transformers/generation/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/generation/utils.py:1286]: Can lower to a warning, and defer this exception if users still fail to detect this after the next release 👍

(No need to change `min_new_tokens`, as it won't be honored when a stopping condition kicks in, like a max length one :) )

[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/generation/utils.py:1286]: Perfect!

**Pull request merged**